### PR TITLE
Singletonize TyInteger/TyFloat

### PR DESCRIPTION
**Pull request by vlad20012**

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt [hunk 1]
```diff
@@ -1112,8 +1112,8 @@ sealed class RsStubLiteralType(val typeOrdinal: Int) {
                     0 -> RsStubLiteralType.Boolean
                     1 -> RsStubLiteralType.Char(readBoolean())
                     2 -> RsStubLiteralType.String(readLong(), readBoolean())
-                    3 -> RsStubLiteralType.Integer(readEnumAsNullable<TyInteger.Kind>())
-                    4 -> RsStubLiteralType.Float(readEnumAsNullable<TyFloat.Kind>())
+                    3 -> RsStubLiteralType.Integer(readByte().toInt().takeIf { it >= 0 }?.let { TyInteger.VALUES[it] })
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/stubs/StubImplementations.kt]: Let's use `List.getOrNull`
I think `TyInteger.VALUES.getOrNull(readByte().toInt())` is more readable than `readByte().toInt().takeIf { it >= 0 }?.let { TyInteger.VALUES[it] }`

**Comment by kumbayo** [comment]
It seems this change causes/exposes a small problem but i am not sure how to best solve.

When i run the single test `RsExpressionTypeInferenceTest`.`test isize in enum variant discriminant` from IntelliJ, i get a `java.lang.NullPointerException`.
If i run all tests in the file (or run the tests via Gradle) the exception disappears.

@vlad20012 Can you also reproduce this problem or is it something that happens on my end only?

From what i can see, the problem is probably related to the initialization order of the companion object and the `ISize` object.
When `TypeInference.infer(element: RsInferenceContextOwner)` executes
`is RsVariantDiscriminant -> TyInteger.ISize to element.expr` this access to `ISize` also causes `VALUES` to be initialized.
But at this time, `ISize` is still `null` which causes a `null` entry in `VALUES`.
Later when `TyInteger.fromSuffixedLiteral` is called, it trips over the unexpected `null` in `VALUES`...

```
java.lang.NullPointerException
	at org.rust.lang.core.types.ty.TyInteger$Companion.fromSuffixedLiteral(TyPrimitive.kt:60)
	at org.rust.lang.core.psi.ext.RsLitExprKt.getStubType(RsLitExpr.kt:24)
	at org.rust.lang.core.types.infer.RsFnInferenceContext.inferLitExprType(TypeInference.kt:697)
	at org.rust.lang.core.types.infer.RsFnInferenceContext.inferType(TypeInference.kt:596)
	at org.rust.lang.core.types.infer.RsFnInferenceContext.inferTypeCoercableTo(TypeInference.kt:622)
	at org.rust.lang.core.types.infer.RsFnInferenceContext.inferLambdaBody(TypeInference.kt:553)
	at org.rust.lang.core.types.infer.RsInferenceContext.infer(TypeInference.kt:139)
	at org.rust.lang.core.types.infer.TypeInferenceKt$inferTypesIn$1.compute(TypeInference.kt:43)
	at org.rust.lang.core.types.infer.TypeInferenceKt$inferTypesIn$1.compute(TypeInference.kt)
	at com.intellij.openapi.util.RecursionManager$2.doPreventingRecursion(RecursionManager.java:112)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:67)
	at org.rust.openapiext.UtilsKt.recursionGuard(utils.kt:44)
	at org.rust.openapiext.UtilsKt.recursionGuard$default(utils.kt:43)
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn(TypeInference.kt:43)
	at org.rust.lang.core.types.ExtensionsKt$inference$1.compute(Extensions.kt:36)
	at com.intellij.psi.util.CachedValuesManager.lambda$getCachedValue$0(CachedValuesManager.java:140)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:49)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:222)
	at com.intellij.psi.impl.PsiCachedValue.getValueWithLock(PsiCachedValue.java:60)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:38)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:87)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:139)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:126)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:35)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:92)
	at org.rust.lang.core.types.ExtensionsKt.getType(Extensions.kt:46)
	at org.rust.lang.core.type.RsTypificationTestBase.check(RsTypificationTestBase.kt:49)
	at org.rust.lang.core.type.RsTypificationTestBase.testExpr(RsTypificationTestBase.kt:23)
	at org.rust.lang.core.type.RsTypificationTestBase.testExpr$default(RsTypificationTestBase.kt:21)
	at org.rust.lang.core.type.RsExpressionTypeInferenceTest.test isize in enum variant discriminant(RsExpressionTypeInferenceTest.kt:426)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at com.intellij.testFramework.UsefulTestCase.lambda$runTest$11(UsefulTestCase.java:339)
	at com.intellij.testFramework.UsefulTestCase.lambda$invokeTestRunnable$12(UsefulTestCase.java:371)
	at com.intellij.testFramework.EdtTestUtilKt.runInEdtAndWait(EdtTestUtil.kt:40)
	at com.intellij.testFramework.UsefulTestCase.invokeTestRunnable(UsefulTestCase.java:370)
	at com.intellij.testFramework.UsefulTestCase.runTest(UsefulTestCase.java:358)
	at com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixtureTestCase.doRunTests(LightPlatformCodeInsightFixtureTestCase.java:99)
	at com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixtureTestCase.runTest(LightPlatformCodeInsightFixtureTestCase.java:90)
	at org.rust.lang.RsTestBase.runTest(RsTestBase.kt:59)
	at com.intellij.testFramework.UsefulTestCase.defaultRunBare(UsefulTestCase.java:384)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:29)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:27)
	at com.intellij.testFramework.EdtTestUtilKt$runInEdtAndWait$2.run(EdtTestUtil.kt:62)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:301)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:346)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by vlad20012** [comment]
@kumbayo I just successfuly reproduced it. Some investigation needed :thinking:

**Comment by vlad20012** [comment]
image

`ISize` is `null` here for a some reason

**Comment by vlad20012** [comment]
Ok, when JVM initializes some class, it request to initialize its parents first. So everything happens exactly as you described. Thanks!
Nice bug. The next question is now to "idiomatically" fix it =D

**Comment by vlad20012** [comment]
:thinking: :thinking: :thinking: 
This seems to work
image

**Comment by kumbayo** [comment]
@vlad20012 Do you have an plan how to solve this? The workaround you posted?
I had this problem happen after startup of the IDEA with certain Rust files.
So it is not only relevant for tests anymore :-(

**Comment by vlad20012** [comment]
@kumbayo, oh, I forget about it, sorry :(
The workaround works. I'll merge it tomorow

**Comment by vlad20012** [comment]
@kumbayo, fixed with #2411

**Pull request merged**